Feature/music main setup

Jowell-E opened the pull request:
-Audio manager added to main scene
-Edited Audiomanager to include a Stop function.  …
-Edited AudioOptions to be serializable
-Added RegisterPlay and RegisterChildren scripts for playing SFX.
-Added PlayAttackVFX script for playing attack animations and sfxs
-Implemented basic music / vfx

Jowell-E committed:
Monster fbx
Damage number indicators
Direction indicator
Slash Effects

Jowell-E committed:
-Damage number indicators
-UI animations
-Renamed PlaceholderCombatUI
-Lighting edits

Jowell-E committed:
Testing line fixed

Jowell-E committed:
Merge branch 'develop' into feature/GameFlow

Jowell-E committed:
Initial setup for music and sfx in main scene

Jowell-E committed:
Merge branch 'develop' of github.com:good-boys/the-good-game into feature/musicMainSetup

# Conflicts:
#	Assets/Audio/SFX.meta
#	Assets/Audio/Temp.meta
#	Assets/Laugh lurker.meta
#	Assets/Scenes/Main.unity
#	Assets/Scripts/Source/Game/Combat/CombatUI.cs

Jowell-E committed:
Audio manager in main scene

Jowell-E committed:
-Edited Audiomanager to include a Stop function.
-Edited AudioOptions to be serializable
-Added RegisterPlay and RegisterChildren scripts for playing SFX.
-Added PlayAttackVFX script for playing attack animations and sfxs
-Implemented basic music / vfx

Jowell-E committed:
-Moved Play VFX function to CombatManager for better control.

imann24 committed:
Add travis_wait to build

imann24 commented on Assets/Scripts/Source/Audio/AudioManager.cs:
```diff
@@ -32,10 +32,61 @@ public void Play(string sourceName, AudioClip clip, AudioOptions options=null)
         {
             options = AudioOptions.GetDefault();
         }
+        if (clip == null)
```
Can this be an overloaded method which does not take a clip as an arg? Or possible call it `Replay`?

imann24 commented on Assets/Scripts/Source/Audio/AudioManager.cs:
```diff
@@ -32,10 +32,61 @@ public void Play(string sourceName, AudioClip clip, AudioOptions options=null)
+        AudioOptions options = clipOptions[getClipIdentifier(source, clip)];
+        Action endHandlerAction = delegate { options.OnEnd(); };
+
+        options.FadeOutTime = fadeOutTime;
```
I'm not sure if it makes sense to permanently overwrite the `AudioOptions` fade out time for this clip. Seems better to keep the local fade out time scoped to the function

imann24 commented on Assets/Scripts/Source/Audio/AudioManager.cs:
```diff
@@ -32,10 +32,61 @@ public void Play(string sourceName, AudioClip clip, AudioOptions options=null)
+            Debug.LogErrorFormat("Source {0} not found. Unable to stop clip.", sourceName);
+            return;
+        }
+        if (clip == null)
```
Overloaded stop method which does not require a clip as an arg?

imann24 commented on Assets/Scripts/Source/Audio/AudioManager.cs:
```diff
@@ -32,10 +32,61 @@ public void Play(string sourceName, AudioClip clip, AudioOptions options=null)
+        Action endHandlerAction = delegate { options.OnEnd(); };
+
+        options.FadeOutTime = fadeOutTime;
+        IEnumerator fadeOutCoroutine = getFadeOutCoroutine(source, options, options.HasEndHandler ? endHandlerAction : null);
```
Nice reuse of this helper method!

imann24 commented on Assets/Scripts/Source/Audio/AudioOptions.cs:
```diff
@@ -29,12 +30,12 @@ public bool HasEndHandler
-    public float FadeOutTime { get; private set; }
-    public float Volume { get; private set; }
-    public bool Looping { get; private set; }
+    public float FadeInTime;
```
Prefer public getter pattern + `[SerializeField]` on private vars (to preserve editor visibility, which was a great addition)!

imann24 commented on Assets/Scripts/Source/Audio/AudioOptions.cs:
```diff
@@ -29,12 +30,12 @@ public bool HasEndHandler
-    public bool HardTransitionIn { get; private set; }
+    public bool HardTransitionIn;
 
     private Action onEndHandler;
```
Maybe worth leveraging `UnityEvent` so we can also set this via the inspector: https://docs.unity3d.com/Manual/UnityEvents.html

imann24 commented on Assets/Scripts/Source/Audio/RegisterChildren.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class RegisterChildren : MonoBehaviour {
```
Nit: rename class to include a mention of the audio system

imann24 commented on Assets/Scripts/Source/Audio/RegisterChildren.cs:
```diff
@@ -0,0 +1,14 @@
+	void Start () {
+        for (int i = 0; i < transform.childCount; i++)
+        {
+            GameFlowManager.instance.audioManager.RegisterAudioSource(transform.GetChild(i).GetComponent<AudioSource>());
```
How about refactoring to: 
```csharp
foreach(AudioSource source in GetComponentsInChildren<AudioSource>())
```

imann24 commented on Assets/Scripts/Source/Audio/RegisterPlay.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class RegisterPlay : MonoBehaviour {
```
Nit: consider an audio specific name. Also, perhaps a way we can combine this logic with `RegisterChildren` via use of parameters, e.g.
```csharp
[SerializeField]
bool registerChildren;

[SerializeField]
bool playOnStart;
```

imann24 commented on Assets/Scripts/Source/Game/GameFlowManager.cs:
```diff
@@ -7,6 +7,7 @@ public class GameFlowManager : MonoBehaviour
     public GameObject camera;
 
     public static GameFlowManager instance = null;
```
😢  http://blog.code-cop.org/2012/01/why-singletons-are-evil.html

imann24 commented on Assets/Scripts/Source/Game/GameFlowManager.cs:
```diff
@@ -7,6 +7,7 @@ public class GameFlowManager : MonoBehaviour
     public GameObject camera;
 
     public static GameFlowManager instance = null;
+    public AudioManager audioManager;
```
Public get, private set?

imann24 commented on Assets/Scripts/Source/Game/Combat/CombatUI.cs:
```diff
@@ -29,7 +29,7 @@ public class CombatUI : AbstractCombatUI
 
     [SerializeField]
-    Image downArrow;
+    Image downArrow;    
```
Nit: whitespace

imann24 commented on Assets/Scripts/Source/Game/Combat/CombatManager.cs:
```diff
@@ -23,6 +23,12 @@ public class CombatManager : MonoBehaviour
     [SerializeField]
     float missTimer = 0.1f;
 
+    [SerializeField]
```
Nit: should there by a helper class which manages VFX ?

imann24 commented on Assets/Scripts/Source/Game/UI/PlayAttackVFX.cs:
```diff
@@ -0,0 +1,95 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+public class PlayAttackVFX : MonoBehaviour {
```
Cool way to encapsulate an attack animation!

imann24 commented on Assets/Scripts/Source/Game/UI/PlayAttackVFX.cs:
```diff
@@ -0,0 +1,95 @@
+
+    public void PlaySFX(string sfxName)
+    {
+        AudioClip targetVFX = null;
```
Nit: var name should be `targetSFX`

imann24 commented on Assets/Scripts/Source/Game/UI/PlayAttackVFX.cs:
```diff
@@ -0,0 +1,95 @@
+        {
+            timer += Time.deltaTime * fadeSpeed;
+
+            newColor.a -= Time.deltaTime;
```
Why are we decreasing the alpha on fade in? 

Also consider using `Color.Lerp` (https://docs.unity3d.com/ScriptReference/Color.Lerp.html)

imann24 commented on Assets/Scripts/Source/Game/Combat/CombatManager.cs:
```diff
@@ -194,6 +203,16 @@ IEnumerator executeActions()
                             }
 
+                            playerHit.PlayVFX("small_0002"); //ToDo: Make this a variable passed by the enemy
+                            if (action.Actor.hitBonus)
```
Ooh! Nice polish

imann24 reviewed (commented):
Some very nice looking audio work!

imann24 committed:
Increase Travis build timeout to 30min

Jowell-E committed:
Audiomanager default no longer has null

Jowell-E committed:
-Renamed registerchildren
-Stop function doesn't overwrite audiooptions and no longer has fadeouttimer arg
-getter / setter stuff cause people like encapsulation or something like that.

imann24 reviewed (approved):
:shipit:
